pybamm/meshes/meshes.py

```diff
@@ -58,14 +62,30 @@ def __init__(self, geometry, submesh_types, var_pts):
                 self.domain_order.append(domain)
 
         for domain in geometry:
-            if "secondary" in geometry[domain].keys():
-                for var in geometry[domain]["secondary"].keys():
-                    repeats = submesh_pts[domain][var.id]  # note (specific to FV)
+            # need to pass tab information if primary domian is 2D current collector
+            if (
+                domain == "current collector"
+                and submesh_types[domain] == pybamm.FenicsMesh2D
+            ):
+                self[domain] = [
+                    submesh_types[domain](
+                        geometry[domain]["primary"],
+                        submesh_pts[domain],
+                        geometry[domain]["tabs"],
+                    )
+                ]
             else:
-                repeats = 1
-            self[domain] = [
-                submesh_types[domain](geometry[domain]["primary"], submesh_pts[domain])
-            ] * repeats
+                if "secondary" in geometry[domain].keys():
+                    repeats = 1
+                    for var in geometry[domain]["secondary"].keys():
+                        repeats *= submesh_pts[domain][var.id]  # note (specific to FV)
+                else:
+                    repeats = 1
+                self[domain] = [
+                    submesh_types[domain](
+                        geometry[domain]["primary"], submesh_pts[domain]
+                    )
+                ] * repeats
```

Comment:
not necessarily linked to this PR but this whole function is pretty opaque and could do with breaking up / more comments